hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreLock.java

```diff
@@ -0,0 +1,533 @@
+                    LockState newState = response.getState();
+                    lockInfo.lockState = newState;
+                    if (newState.equals(LockState.WAITING)) {
+                      throw new WaitingForLockException(
+                          String.format(
+                              "Waiting for lock on table %s.%s", databaseName, tableName));
+                    }
+                  } catch (InterruptedException e) {
+                    Thread.interrupted(); // Clear the interrupt status flag
+                    LOG.warn(
+                        "Interrupted while waiting for lock on table {}.{}",
+                        databaseName,
+                        tableName,
+                        e);
+                  }
+                },
+                TException.class);
+      }
+    } catch (WaitingForLockException e) {
+      timeout = true;
+      duration = System.currentTimeMillis() - start;
+    } catch (TException e) {
+      error = e;
+    } finally {
+      if (!lockInfo.lockState.equals(LockState.ACQUIRED)) {
+        unlock(Optional.of(lockInfo.lockId));
+      }
+    }
+
+    if (lockInfo.lockState.equals(LockState.ACQUIRED)) {
+      return lockInfo.lockId;
+    }
+
+    // timeout and do not have lock acquired
+    if (timeout) {
+      throw new LockException(
+          "Timed out after %s ms waiting for lock on %s.%s", duration, databaseName, tableName);
+    }
+
+    // On thrift error and do not have lock acquired
+    if (error != null) {
+      throw new LockException(
+          error, "Metastore operation failed for %s.%s", databaseName, tableName);
+    }
+
+    // Just for safety. We should not get here.
+    throw new LockException(
+        "Could not acquire the lock on %s.%s, lock request ended in state %s",
+        databaseName, tableName, lockInfo.lockState);
+  }
+
+  /**
+   * Tries to create a lock. If the lock creation fails, and it is possible then retries the lock
+   * creation a few times. If the lock creation is successful then a {@link LockInfo} is returned,
+   * otherwise an appropriate exception is thrown.
+   *
+   * @return The created lock
+   * @throws LockException When we are not able to fill the hostname for lock creation, or there is
+   *     an error during lock creation
+   */
+  @SuppressWarnings("ReverseDnsLookup")
+  private LockInfo tryLock() throws LockException {
+    LockInfo lockInfo = new LockInfo();
+
+    String hostName;
+    try {
+      hostName = InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException uhe) {
+      throw new LockException(uhe, "Error generating host name");
+    }
+
+    final LockComponent lockComponent =
+        new LockComponent(LockType.EXCLUSIVE, LockLevel.TABLE, databaseName);
+    lockComponent.setTablename(tableName);
+    final LockRequest lockRequest =
+        new LockRequest(
+            Lists.newArrayList(lockComponent), System.getProperty("user.name"), hostName);
```

Comment:
Just FYI, I'm trying to promote the usage of Hadoop's UserGroupInformation instead of OS user when interacting with HMS. In production Hive environment where Hadoop authN mechanism is enabled, the Hive ecosystem will not integrate well with the local OS user name (they typically use Kerberos principal)

Details at: https://github.com/apache/iceberg/pull/6324/files#diff-4f691cd4a9db7de8859c76a837a56fd507137d5a1570a0cfed3db6fca41c0f19R32

Reply:
Done in #6648